File: Data Structure/Union find/Hard/LET_no952_Largest Component Size by Common Factor.cpp
class unionFind {
private:
    vector<int>parent, size, rank;
public:
    unionFind(vector<int>& arr, int n) {
        parent.resize(n);
        size.resize(n);
        rank.resize(n);
        for (int i = 0; i < n; i++)
            parent[i] = i;
        for (int num : arr)
            size[num] = 1;
    }
    
    int find(int x) {
        if (parent[x] != x)
            parent[x] = find(parent[x]);
        return parent[x];
    }
    
    void Union(int x, int y) {
        int px = find(x), py = find(y);
        if (px != py) {
            if (rank[px] < rank[py]) {
                parent[px] = py;
                size[py] += size[px];
                size[px] = 0;
            }
            else {
                parent[py] = px;
                size[px] += size[py];
                size[py] = 0;
                rank[px] += rank[px] == rank[py];
            }
        }
    }
    
    int largestComponentSize() {
        return *max_element(size.begin(), size.end());
    }
};

class Solution {
public:
    int largestComponentSize(vector<int>& nums) {
        int n = nums.size(), mx = *max_element(nums.begin(), nums.end());
        unionFind uf(nums, mx + 1);
        
        for (int num : nums) {
            for (int i = 2; i * i <= num; i++) {
                if (num % i == 0) {
                    uf.Union(num / i, num);
                    uf.Union(i, num);
                }
            }
        }
        
        return uf.largestComponentSize();
    }
};

File: Data Structure/Union find/Hard/LET_no952_Largest Component Size by Common Factor.py
class unionFind:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
    
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def union(self, x, y):
        px, py = self.find(x), self.find(y)
        self.parent[px] = py

class Solution:
    def largestComponentSize(self, nums: List[int]) -> int:
        n = len(nums)
        uf = unionFind(n)
        primes = collections.defaultdict(list)
        for i, num in enumerate(nums):
            for p in self.getPrimes(num):
                primes[p].append(i)
        
        for _, indexes in primes.items():
            for i in range(len(indexes) - 1):
                uf.union(indexes[i], indexes[i + 1])
                
        return max(Counter([uf.find(i) for i in range(n)]).values())    
    
    def getPrimes(self, x):
        for i in range(2, int(math.sqrt(x)) + 1):
            if x % i == 0:
                return self.getPrimes(x // i) | set([i])
        return set([x])